### ISSUE TITLE
Could not find root view for a given ancestor with tag

### ISSUE THREAD

**Issue by federicot** [kmagiera/react-native-gesture-handler#1675]
## Description

We are seeing the following crash on Android:
```
com.facebook.react.bridge.JSApplicationIllegalArgumentException: Could find root view for a given ancestor with tag 3199
    at com.swmansion.gesturehandler.react.RNGestureHandlerModule.tryInitializeHandlerForReactRootView(RNGestureHandlerModule.java:18)
    at com.swmansion.gesturehandler.react.RNGestureHandlerModule.attachGestureHandler(RNGestureHandlerModule.java:1)
    at java.lang.reflect.Method.invoke(Method.java)
    at com.facebook.react.bridge.JavaMethodWrapper.invoke(JavaMethodWrapper.java:18)
    at com.facebook.react.bridge.JavaModuleWrapper.invoke(JavaModuleWrapper.java:2)
    at com.facebook.react.bridge.queue.NativeRunnable.run(NativeRunnable.java)
    at android.os.Handler.handleCallback(Handler.java:869)
    at android.os.Handler.dispatchMessage(Handler.java:101)
    at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(MessageQueueThreadHandler.java:1)
    at android.os.Looper.loop(Looper.java:206)
    at com.facebook.react.bridge.queue.MessageQueueThreadImpl$4.run(MessageQueueThreadImpl.java:8)
    at java.lang.Thread.run(Thread.java:764)
```

### Screenshots
<img width="404" alt="Screenshot 2021-10-04 at 17 09 33" src="https://user-images.githubusercontent.com/681864/136389518-05d39127-8c7e-41a4-8599-43db0ebba137.png">

## Steps To Reproduce

I was not able to reproduce it consistently, but it seems to happen after the app goes to foreground.

## Package versions

- React: 17.0.1
- React Native: >= 0.64.2
- React Native Gesture Handler: 1.10.3
- React Navigation: 5.14.4

**Comment by rdeift** [comment]
+1

**Comment by ThienNgn** [comment]
+1

**Comment by systemride** [comment]
We are also seeing this with:
```
"react-native": "0.66.1",
"react-native-gesture-handler": "1.8.0",
```

**Comment by aleksandr-senichev** [comment]
+1

**Comment by LFMAKER** [comment]
+1

**Comment by stephenkopylov** [comment]
+1
```
com.facebook.react.bridge.JSApplicationIllegalArgumentException: Could find root view for a given ancestor with tag 6343
        at com.swmansion.gesturehandler.react.RNGestureHandlerModule.tryInitializeHandlerForReactRootView(RNGestureHandlerModule.java:18)
        at com.swmansion.gesturehandler.react.RNGestureHandlerModule.attachGestureHandler(RNGestureHandlerModule.java:1)
        at java.lang.reflect.Method.invoke(Method.java:-2)
        at com.facebook.react.bridge.JavaMethodWrapper.invoke(JavaMethodWrapper.java:18)
        at com.facebook.react.bridge.JavaModuleWrapper.invoke(JavaModuleWrapper.java:13)
        at com.facebook.react.bridge.queue.NativeRunnable.run(NativeRunnable.java:-2)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(MessageQueueThreadHandler.java:1)
        at android.os.Looper.loop(Looper.java:214)
        at com.facebook.react.bridge.queue.MessageQueueThreadImpl$4.run(MessageQueueThreadImpl.java:8)
        at java.lang.Thread.run(Thread.java:919)
```

**Comment by anatooly** [comment]
Same issue here too
"react-native": "0.66.1",
"react-native-gesture-handler": "^1.10.3",

**Comment by systemride** [comment]
This may not really be a fix, but we found a workaround for this that is working well for us. We are using:

```
"react-native": "0.66.1",
"react-native-gesture-handler": "1.8.0",
```

1. Remove the Override in MainActivity.java:
```
-  @Override
-  protected ReactActivityDelegate createReactActivityDelegate() {
-    return new ReactActivityDelegate(this, getMainComponentName()) {
-      @Override
-      protected ReactRootView createRootView() {
-       return new RNGestureHandlerEnabledRootView(MainActivity.this);
-      }
-    };
-  }
```

2. Add the GestureHandlerRootView to your app.
```
+ import { GestureHandlerRootView } from "react-native-gesture-handler";

...

+ <GestureHandlerRootView>
+   <App/>
+ </GestureHandlerRootView>
```

3. This step might not be necessary and probably not recommended by the library authors, but we also forked our own version and made the following changes to `RNGestureHandlerModule.java`:
```
- throw new JSApplicationIllegalArgumentException("Could find root view for a given ancestor with tag "
              + ancestorViewTag);
```

... and just return instead of throwing. Honestly we're not exactly sure of the ramifications of this, but the app seems to behave normally in this scenario.
```
+ return;
```

**Comment by federicot** [comment]
@systemride Thanks for sharing. Are you also using codepush?

**Comment by yemd** [comment]
> This may not really be a fix, but we found a workaround for this that is working well for us. We are using:
> 
> ```
> "react-native": "0.66.1",
> "react-native-gesture-handler": "1.8.0",
> ```
> 
> 1. Remove the Override in MainActivity.java:
> 
> ```
> -  @Override
> -  protected ReactActivityDelegate createReactActivityDelegate() {
> -    return new ReactActivityDelegate(this, getMainComponentName()) {
> -      @Override
> -      protected ReactRootView createRootView() {
> -       return new RNGestureHandlerEnabledRootView(MainActivity.this);
> -      }
> -    };
> -  }
> ```
> 
> 2. Add the GestureHandlerRootView to your app.
> 
> ```
> + import { GestureHandlerRootView } from "react-native-gesture-handler";
> 
> ...
> 
> + <GestureHandlerRootView>
> +   <App/>
> + </GestureHandlerRootView>
> ```
> 
> 3. This step might not be necessary and probably not recommended by the library authors, but we also forked our own version and made the following changes to `RNGestureHandlerModule.java`:
> 
> ```
> - throw new JSApplicationIllegalArgumentException("Could find root view for a given ancestor with tag "
>               + ancestorViewTag);
> ```
> 
> ... and just return instead of throwing. Honestly we're not exactly sure of the ramifications of this, but the app seems to behave normally in this scenario.
> 
> ```
> + return;
> ```

Well, unfortunately, your trick won't do for me, my app goes completely blank

**Comment by systemride** [comment]
> Well, unfortunately, your trick won't do for me, my app goes completely blank

I updated my comment, you might also try adding flex styles to the component. Eg:
```
<GestureHandlerRootView style={{ flex: 1 }}>
```

**Comment by systemride** [comment]
> @systemride Thanks for sharing. Are you also using codepush?

No, not using codepush.

**Comment by yemd** [comment]
> 

Damn, I did not every thought about that, I ended up using patch-package to patch the module and return instead of throw, I don't have any overrides on my MainActivity.java, it's working fine, I may try your way tho thanks!

**Comment by nchase** [comment]
@federicot You asked if another user was using Codepush. 

Are you aware of a relationship between Codepush and this crash?

My team is seeing an increase in crashes with this stacktrace, and it's not clear to us why.  Unlike @systemride , we _are_ using Codepush, and the crashes seem to have increased after a Codepush.

**Comment by federicot** [comment]
> @federicot You asked if another user was using Codepush.
> 
> Are you aware of a relationship between Codepush and this crash?
> 
> My team is seeing an increase in crashes with this stacktrace, and it's not clear to us why. Unlike @systemride , we _are_ using Codepush, and the crashes seem to have increased after a Codepush.

@nchase We are seeing a slight correlation when we do codepush releases. There's also this related issue https://github.com/software-mansion/react-native-gesture-handler/issues/1724

**Comment by stephenkopylov** [comment]
We use codepush too

**Comment by aleksandr-senichev** [comment]
As a temporal solution, we made a fork of stack navigator and removed the `GestureHandlerRootView` wrapper for the Android app, because it doesn't use any gesture feature in our case. It works stable now.

**Comment by jayshah-swiggy** [comment]
Codepush user, also seeing similar crash

**Comment by aamagda** [comment]
We have the same issue after the codepush releases

**Comment by aamagda** [comment]
There is interesting analysis by @netshade
https://github.com/software-mansion/react-native-gesture-handler/issues/715#issuecomment-816899347

@kmagiera Hi, could you explain please the reason for checking `rootViewTag < 1` and not for `rootViewTag == View.NO_ID`   inside [tryInitializeHandlerForReactRootView(ancestorViewTag: Int)](https://github.com/software-mansion/react-native-gesture-handler/blob/74ef813fa3a92bd3947ea9f246468fc1dd0a9f34/android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.kt#L455) method?

Do you have a plan to fix this issue or maybe any workarounds? What do you think about this advice https://github.com/software-mansion/react-native-gesture-handler/issues/1675#issuecomment-979411493

**Comment by jbaudanza** [comment]
I was able to resolve this issue in my project. The problem is that `Codepush.restartApp` doesn't actually restart the app. Instead, it uses the ReactNativeInstanceManager class to reload the bundle:

https://github.com/facebook/react-native/blob/4fb49cd555942caba1515b6a03f8ccea931dfc90/ReactAndroid/src/main/java/com/facebook/react/ReactInstanceManager.java#L435

This means that if any RN-GestureHandler components begin initializing before a `Codepush.restartApp`, there is a very good chance that when they finish., the RootNode doesn't exist anymore. This is what causes the crash.

To solve this, you have to make sure there are no instances of RN-GestureHandler in the process of loading before a `Codepush.restartApp` operation. This includes things like `react-navigation`, which will mount gesture handlers behind the scenes.

In our case, when our app loads we begin a codepush sync. While the codepush is in any of these states, we render a simple View with no gesture handlers or react-navigation are mounted: `CHECKING_FOR_UPDATE`, `DOWNLOADING_PACKAGE`, `UPDATE_INSTALLED`, `INSTALLING_UPDATE`.

Only after the codepush is completely finished, it's safe to mount any react-navigation components.

**Comment by scesbron** [comment]
I have the same problem with code push in production. What I don't understand is that it is not systematic and in staging we are not able to reproduce it.

**Comment by jbaudanza** [comment]
> I have the same problem with code push in production. What I don't understand is that it is not systematic and in staging we are not able to reproduce it.

It's not systematic because it's a race condition between 'attachGestureHandler' and codepush/ReactNativeInstanceHandler. If the gesture handler wins, then there is no crash. If you try it on an older Android phone, you should be able to reproduce it. I have a Samsung galaxy J3 and can reproduce it fairly often.

**Comment by nchase** [comment]
Does anyone have steps that we can follow to consistently reproduce this bug? (ideally in an emulator)

**Comment by thekingrenz23** [comment]
+1

**Comment by tijs** [comment]
We have a lot of these crashes but we do not use codepush so while it seems to be triggered by a codepush restart it can also happen in apps that do not use codepush at all.

**Comment by BPatinho** [comment]
We're facing the same issue with OTA updates by expo-updates. It's happening since we upgraded react native to 0.66

**Comment by kperdomo1** [comment]
we found a way to test this out. In our case, our codePush installMode uses the `ON_NEXT_RESUME` strategy, so after downloading the codePush update it is installed when user comes back to the foreground.

1. Launch an android emulator with low resources (in our case, 343 MB of RAM, 48PB of VM heap, 2 multi-core CPU)
2. Find a way to trigger attaching a new gesture handler. In our case, it was enough to navigate to other screen when the app state changes from `background` to `active`.
3. Release an update to CodePush so you can download it in your app.
3. Launch your app and wait for the CodePush update to be downloaded
4. Set app to the background
5. Come back to the foreground
6. At this point the race condition pointed out by @jbaudanza should happen and the app crashes.

Hope this helps

**Comment by princeofv** [comment]
![image](https://user-images.githubusercontent.com/47618675/151905176-e9193da9-eb42-4514-8a03-0d8ec9c74e3c.png)

**Comment by nchase** [comment]
@yemd have you discovered any issues related to your workaround?

**Comment by systemride** [comment]
> @systemride have you discovered any issues related to your workaround?

So far we have not discovered any other issues to our workaround but it's still something we want to revisit. 

Also maybe worth noting... that while we don't use codepush, we do use [react-native-restart](https://github.com/avishayil/react-native-restart) which might do something similar to codepush.

**Comment by LFMAKER** [comment]
+1

**Comment by johnsoncwb** [comment]
same problem here, any solution?

**Comment by nchase** [comment]
@kmagiera maybe this is a question that you can answer, since it seems that you may have written this code originally: why does this library [throw an error in this case](https://github.com/software-mansion/react-native-gesture-handler/blob/9c7664b7fa426f366c68155181082bcf52943125/android/src/main/java/com/swmansion/gesturehandler/react/RNGestureHandlerModule.java#L579-L582)?

**Comment by deboralbarros** [comment]
We have the same issue here, any updates?

**Comment by LFMAKER** [comment]
We're facing the same issue with OTA updates by expo-updates. It's happening since we upgraded react native to 0.66
@j-piasecki @kmagiera

**Comment by itenl** [comment]
0.66.2 + react-native-gesture-handler + CodePush same issues

**Comment by newbanker-tu** [comment]
> > @federicot You asked if another user was using Codepush.
> > Are you aware of a relationship between Codepush and this crash?
> > My team is seeing an increase in crashes with this stacktrace, and it's not clear to us why. Unlike @systemride , we _are_ using Codepush, and the crashes seem to have increased after a Codepush.
> 
> Our team is also using Codepush. Crashes with this stack trace is pretty common after a codepush restart on Android. We have our app wrapped in `GestureHandlerRootView`, but maybe something about the codepush restart causes this to become unmounted? I'm planning on looking more into it this week.

seem to you，did you resolve？

**Comment by LFMAKER** [comment]
@j-piasecki This PR #1936 doesn't fix it. I have a production version with 2.4.0 and the error continues.
<img width="1118" alt="Screen Shot 2022-04-20 at 09 53 30" src="https://user-images.githubusercontent.com/8159987/164234859-99d791e4-60ad-4f16-8cbc-917535fcbdcb.png">

**Comment by abdymm** [comment]
hi @LFMAKER @brduck are you guys find a way to solve this on expo?

**Comment by xwartz** [comment]
are there any solutions?

**Comment by sebas21** [comment]
any solution??? RN 0.69

**Comment by VashaMasha** [comment]
> This may not really be a fix, but we found a workaround for this that is working well for us. We are using:
> 
> ```
> "react-native": "0.66.1",
> "react-native-gesture-handler": "1.8.0",
> ```
> 
>  1. Remove the Override in MainActivity.java:
> 
> ```
> -  @Override
> -  protected ReactActivityDelegate createReactActivityDelegate() {
> -    return new ReactActivityDelegate(this, getMainComponentName()) {
> -      @Override
> -      protected ReactRootView createRootView() {
> -       return new RNGestureHandlerEnabledRootView(MainActivity.this);
> -      }
> -    };
> -  }
> ```
> 
>  2. Add the GestureHandlerRootView to your app.
> 
> ```
> + import { GestureHandlerRootView } from "react-native-gesture-handler";
> 
> ...
> 
> + <GestureHandlerRootView style={{ flex: 1 }}>
> +   <App/>
> + </GestureHandlerRootView>
> ```
> 
>  3. This step might not be necessary and probably not recommended by the library authors, but we also forked our own version and made the following changes to `RNGestureHandlerModule.java`:
> 
> ```
> - throw new JSApplicationIllegalArgumentException("Could find root view for a given ancestor with tag "
>               + ancestorViewTag);
> ```
> 
> ... and just return instead of throwing. Honestly we're not exactly sure of the ramifications of this, but the app seems to behave normally in this scenario.
> 
> ```
> + return;
> ```

Thank you!
Worked for me

"react-native": "0.67.2",
"react-native-gesture-handler": "2.2.0",

Version numbers stayed the same and I don't have the code from step 3. So I only implemented steps 1 and 2
